[5.4] Fluent partial resource routes

jaripekkala opened the pull request:
Discussed in https://github.com/laravel/internals/issues/412

In 5.4 you can register routes more fluently e.g.

```php
Route::name('users.index')->get('users', 'UserController@index');
```

It would be nice to get same behaviour for registering partial resource routes e.g. 

```php
Route::resource('users', 'UserController')->only('index', 'show');

Route::resource('comments', 'CommentController')->except('destroy');

Route::resource('posts', 'PostController')->only(
    'index', 'show', 'store', 'update', 'destroy'
);
```

Commit:
add lazy registering resources tests

Commit:
return PendingResourceRegistration when
registering resources which registeres them on destruction possibly limiting/excluding methods

JosephSilber commented:
This relies heavily on GC.

Is it guaranteed to run before the actual route is matched? Or before the routes are cached?

GrahamCampbell commented:
@JosephSilber is correct. This implement is not guaranteed to be correct.

GrahamCampbell commented:
👎

Commit:
add method to remove existing routes
from the collection

Commit:
register resource routes immediately
allow limiting and excluding routes afterwards

Commit:
refactor

compact the resource data into array and extract getting the excluded methods to their own methods

salomoni commented:
Implemented now in the other way proposed https://github.com/laravel/internals/issues/412#issuecomment-280303057

Now `only` and `except` removes registered routes from the `RouteCollection`.

Commit:
fix route removing undefined index

taylorotwell commented:
If the PHP manual is correct, then yes the implementation is guranteed to be correct:

"The destructor method will be called **as soon as there are no other references to a particular object**"

JosephSilber commented:
I doubt that is to be taken so literal.

There's no way for the runtime to know that there are no longer any references to an object. The only way to know that is at the end of a GC cycle.

GrahamCampbell commented:
I don't think the PHP manual is correct. It is up to the garbage collector to look for objects with a zero refcount.

taylorotwell commented:
So you're going to submit a PR to the PHP documentation to fix it? It states it very specifically in several places :)

taylorotwell commented:
And if this is the case, should we take ANYTHING in the PHP manual literally? If so, why?

browner12 commented:
I'll add my two cents from personal experience, I've had less than stellar results when dealing with destructors. It seemed to be very difficult to rely on when it was run.

taylorotwell commented:
I would actually be very interested if someone can show a situation where it is not run as soon as there are no other references to the object.

GrahamCampbell commented:
> I would actually be very interested if someone can show a situation where it is not run as soon as there are no other references to the object like the documentation states.

By its very nature, it's non-deterministic, so, it's hard to provide an example and make any guarantees about when garbage collection will take place.

GrahamCampbell commented:
That said, I was unable to get PHP to call these destructors in an "incorrect" order: https://3v4l.org/qUD51. Maybe there's an ordering by the garbage collector, just due to the implementation specifics?

GrahamCampbell commented:
There are definitely user error cases I can think of though. Will write an example.

GrahamCampbell commented:
Here we go: https://3v4l.org/ihpuf.

The application runs to completion, before the route gets registered,

GrahamCampbell commented:
In terms of the PR, I actually already have a big problem with it, the `register` method is not supposed to return anything, but now it does...

(and if the programmer actually leaves a reference to the returned value somewhere, they are screwed)

GrahamCampbell commented:
For that reason, I think this should be a 5.5 change, because it changes the contract.

GrahamCampbell commented:
An implementation to avoid using destructors, would be for the registra to keep a queue of pending routes, and call some resolve method on them all, before we try to match anything.

jaripekkala committed:
Revert "fix route removing undefined index"

This reverts commit 713483d97e06b7ef172c28d9048afd0196d557b9.

jaripekkala committed:
Revert "refactor"

This reverts commit fb6afbb8f8c27a7e68f85dda08709a93ce07711b.

jaripekkala committed:
Revert "register resource routes immediately"

This reverts commit 48336ab990185b7aa024e2cad784a5518af1b51b.

jaripekkala committed:
Revert "add method to remove existing routes"

This reverts commit 4254df357db81b1ad1d72e7d5bab88e6ce00d032.

jaripekkala committed:
rename destruct to register,
combine resource info to one array instead of storing each in props

jaripekkala committed:
rename lazy to registration, forget remember metho

jaripekkala committed:
resource method now PendingResourceRegistration

jaripekkala committed:
hold the resources pending until needed

jaripekkala committed:
don't hold the prefixed resources

salomoni commented:
Thank you @GrahamCampbell for your time!

I implemented this once again based on your suggestion.

taylorotwell commented:
This really was cleaner using destructors tbh, even though some are scared they aren't called when the PHP docs say they are called. Graham's code example is not really reflective of an actual Laravel routes file.

GrahamCampbell commented:
> Graham's code example is not really reflective of an actual Laravel routes file.

There's no reason to assume that people do routing in that way. I don't always.

GrahamCampbell commented:
This solution is robust. And I don't think it's any worse than the destructor way tbh.

GrahamCampbell commented on src/Illuminate/Contracts/Routing/Registrar.php:
```diff
@@ -74,7 +74,7 @@ public function match($methods, $uri, $action);
      * @param  string  $controller
      * @param  array   $options
-     * @return void
+     * @return \Illuminate\Routing\PendingResourceRegistration
```
this PR should target 5.5, because this is a breaking change to the interface, requiring implementations to return this object

salomoni replied:
A newbie question: does this require any actions by me? My first PR 🥇

taylorotwell commented:
Is this not changing the entire order of when your routes are registered though? Now the resource routes aren't *actually* registered until a later time and that makes a difference when registering things.

taylorotwell commented:
Also I still haven't seen an actual case that won't work with the destructor approach. 😬

GrahamCampbell commented:
> Also I still haven't seen an actual case that won't work with the destructor approach.

Yeh, u have. :trollface: That example where I showed the variable living beyond the scope we wanted it to. I mean, I could write a laravel app that demonstrates it, but I can't really be arsed.

The point is, that if a route remains referenced, after routing happens, then it won't have ever been registered. It's likely people would do this by mistake, but it's very hard for them to debug, if they don't understand the internals.

JosephSilber commented:
The point here is more than that. Even if you don't have a reference to it anymore, there's no guarantee when GC will run.

PHP's docs be damned.

GrahamCampbell commented:
> there's no guarantee when GC will run.

While that is true, I haven't been to actually get PHP to not garbage collect in time in my testing to try to break it.

JosephSilber commented:
@GrahamCampbell that's not enough to actually build a feature on top of.

GrahamCampbell commented:
> that's not enough to actually build a feature on top of.

I agree, that's why I asked for this to be implemented without using that, and it now has been.

jaripekkala committed:
testCanNotLimitMethodsAfterRegistered

salomoni commented:
The resources are currently registered only once so calling `only` or `except` methods after registration has no effect.

Should they?

Pending resources could just be marked as registered and exception thrown if tried to limit/exclude methods afterwards.

Registering resources in the RouteCollection each time needed is a little overkill imo. There might e.g. be several `route('users.index')` calls  in the templates.

taylorotwell commented:
If you want to re-submit this using the original, simple approach I suggested go for it. I can't really subscribe to the view that the PHP documentation is not to be trusted because then we can't program anything in this language at all to be honest. I will function as if the documentation is true. If it is not, it should be fixed.

GrahamCampbell commented:
For what it's worth 👎 On the original approach. This new approach is much better IMO. (ultimately, do what taylor says tho, he da boss)
